daft/datatype.py

```diff
@@ -131,6 +125,20 @@ def struct(cls, fields: dict[str, DataType]) -> DataType:
     def extension(cls, name: str, storage_dtype: DataType, metadata: str | None = None) -> DataType:
         return cls._from_pydatatype(PyDataType.extension(name, storage_dtype._dtype, metadata))
 
+    @classmethod
+    def embedding(cls, name: str, dtype: DataType, size: int) -> DataType:
+        if not isinstance(size, int) or size <= 0:
+            raise ValueError("The size for a embedding must be a positive integer, but got: ", size)
+        return cls._from_pydatatype(PyDataType.embedding(name, dtype._dtype, size))
+
+    @classmethod
+    def image(cls, name: str, dtype: DataType, shape: tuple | None = None) -> DataType:
```

Comment:
should we fixed to H, W, C here rather than a tuple to ensure consistency when writing our kernels

Reply:
Are you thinking:
1. kwargs for each - `daft.image("im", daft.float32(), h=28, w=28, c=4)`
2. `size` and `num_channels` kwargs - `daft.image("im", daft.float32(), size=(28, 28), num_channels=3)`
3. Make `shape` a 3-element tuple with the semantics given in the docstring and maybe the arg name- `daft.image("im", daft.float32(), hwc_shape=(28, 28, 3)`
4. Something else?

Reply:
Actually thinking about it more, what do you think about having an `ImageMode` enum. Like `RGB`, `BGR`, `GrayScale`, `RBGA` and so forth.

Reply:
That would be in addition just having `H x W`

Reply:
So something closer to the PIL paradigm? I stayed away from an image mode enum since I thought that mode handling would delay this first PR, but it actually doesn't seem too bad now that I think about it. Let me quickly add that!

Reply:
Also should the size (`H x W`) be optional but have the mode always be required, or allow the mode to be optional as well? I'm thinking that the mode should always be required.

Reply:
Also also, PIL encodes the underlying dtype in the mode, so if we make the mode required, we can eliminate the dtype arg.

Reply:
I think for the VariableSizedImage, the mode should be encoded per image. Example being that you're loading a bunch of images from cloud storage that have mixed modes.

There's some other things to consider.

1. Do we want to allow the dtype and mode to be independent (lets say I wanted to normalize an RGB image, from uint8 to float32)
2.  Do we want to say that any of those operations would return a TensorType instead rather than an image?

My understanding is that PIL doesn't support non-mode dtypes, which has been annoying in the past when you want run a crop or resize on a matrix of floats

Reply:
Ah, just seeing your most recent comment now after I pushed up a version with the mode being top-level, on the type.

>My understanding is that PIL doesn't support non-mode dtypes, which has been annoying in the past when you want run a crop or resize on a matrix of floats

Yep I know that's been a source of annoyance for users in the past, so having a better story there than PIL would be great.

>Do we want to allow the dtype and mode to be independent (lets say I wanted to normalize an RGB image, from uint8 to float32)

I think we have a few options here:
1. **Mode required, dtype overridable** - we make the dtype inferred from the mode by default, and then make it overridable with an optional `dtype` arg.
2. **Mode optional, dtype optional** - must specify at least one; if both are specified, the specified dtype takes precedence over the dtype inferred from the mode.
3. **Both mode and dtype required** - mode only determines the number of channels and some basic semantics, and is orthogonal to the dtype (we wouldn't support the PIL image modes that are only concerned with dtypes).

>Do we want to say that any of those operations would return a TensorType instead rather than an image?

I think this comes down to whether we want to support image-specific operations for images that deviate from our mode scheme, or if tensor operations would suffice.

> I think for the VariableSizedImage, the mode should be encoded per image. Example being that you're loading a bunch of images from cloud storage that have mixed modes.

We can certainly do that, although we should probably then store the mode and size separately rather than translating the mode + size into an overall image shape (including the channel dimension).